Add chrome://rewards-internals to assist with support

emerick opened the pull request:
Fixes brave/brave-browser#1174

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [ ] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [x] Windows
  - [ ] macOS
  - [ ] Linux
- Verified that these changes pass automated tests (`npm test brave_unit_tests && npm test brave_browser_tests`) on
  - [x] Windows
  - [ ] macOS
  - [ ] Linux
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [ ] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [x] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone

## Test Plan:

- Start Brave
- Navigate to chrome://version
- Click the "More details about Brave Rewards" link to go to chrome://rewards-internals
- Verify that information looks correct

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

emerick commented:
This is how the chrome://version page currently looks with the new link. Other suggestions welcome, though.

![brave-rewards-details-link](https://user-images.githubusercontent.com/5033691/50234170-0cb07200-0383-11e9-9afa-aa5334da8c5e.png)

emerick commented:
chrome://rewards-internals looks like this currently:

![image](https://user-images.githubusercontent.com/5033691/50234420-ac6e0000-0383-11e9-8d0c-dc36683e235d.png)

A bit bare, looking for feedback on what else we can add here...

NejcZdovc commented:
@emerick can you please fix merge conflicts

NejcZdovc commented on components/brave_rewards/resources/internals/components/app.tsx:
```diff
@@ -0,0 +1,41 @@
+      <div id='rewardsInternalsPage'>
+        <KeyInfoSeed keyInfoSeed={rewardsInternalsData.info.keyInfoSeed || ''} />
+        <WalletPaymentId walletPaymentId={rewardsInternalsData.info.walletPaymentId || ''} />
+	<hr/>
```
alignment is off here

NejcZdovc commented on components/brave_rewards/resources/internals/components/app.tsx:
```diff
@@ -0,0 +1,41 @@
+	<hr/>
+	{rewardsInternalsData.info.currentReconciles.map((item, index) => (
+	    <div>
+	    <span i18n-content='currentReconcile'/> {index + 1}
```
please indent

NejcZdovc commented on components/common/typescript.gni:
```diff
@@ -152,4 +153,4 @@ template("pack_web_resources") {
   }
-}
\ No newline at end of file
+}
```
do we need new line here?

bridiver replied:
I think this is just standard eof newline. My editor automatically adds it and his probably does too

NejcZdovc commented on components/definitions/rewardsInternals.d.ts:
```diff
@@ -0,0 +1,20 @@
+  }
+
+  export interface State {
+    info: {
```
nit: maybe just have one level? (so that we don't need info wrapper)

emerick replied:
Makes sense

NejcZdovc commented on components/services/bat_ledger/bat_ledger_impl.h:
```diff
@@ -120,8 +120,10 @@ class BatLedgerImpl : public mojom::BatLedger,
+        GetRewardsInternalsInfoCallback callback) override;
 
-  private:
+   private:
```
I think this one was not intentional

emerick commented on vendor/bat-native-ledger/src/ledger_impl.cc:
```diff
@@ -1006,6 +1006,30 @@ void LedgerImpl::SetWalletInfo(
+  const braveledger_bat_helper::WALLET_INFO_ST wallet_info =
+      bat_state_->GetWalletInfo();
+  if (wallet_info.keyInfoSeed_.size() > 0)
+    info.key_info_seed = braveledger_bat_helper::getBase64(wallet_info.keyInfoSeed_);
```
OK, makes sense. What do you recommend as far as determining if the key info seed is valid? I can check if the seed is `SEED_LENGTH` bytes in length and then check if `getPublicKeyFromSeed` is successful. Does that seem sufficient or is there a better approach?

NejcZdovc replied:
this sounds good to me yes

emerick commented on vendor/bat-native-ledger/include/bat/ledger/rewards_internals_info.h:
```diff
@@ -0,0 +1,33 @@
+  const std::string ToJson() const;
+  bool loadFromJson(const std::string& json);
+
+  struct CurrentReconcileInfo {
```
Makes sense, will do.

emerick commented:
@NejcZdovc This is ready for review again. I think I addressed all of your concerns, but let me know.

NejcZdovc commented:
1) can you please show something like "Rewards are not enabled. Please enabled them [here](brave://rewards/)." when rewards are not enabled and you go to brave://rewards-internals/ before enabling rewards. And other info is hidden until you enable rewards.

2) If I corrupt key info seed, wallet payment ID is empty. We should still show it
![image](https://user-images.githubusercontent.com/9574457/53233049-b3b7cf80-368c-11e9-8f2b-e6aa1dfac206.png)

emerick commented:
I think what's actually happening in #2 is that the ledger state hasn't finished loading at the time the chrome://rewards-internals page is displayed. This can happen if you have Brave set to save your tabs on exit, visit chrome://rewards-internals, exit Brave and then restart it (the race condition is much more likely in that scenario).

Do we have an existing notification exposed from Rewards that lets us know when rewards is fully initialized? If not, sounds like I'll need to add one and then listen for it before showing the internals info.

emerick commented:
Probably something like `OnRewardsInitialized` or `OnRewardsReady`...

emerick commented:
@NejcZdovc Ready for review again!

emerick commented:
Updated patches for C73.

bridiver commented on components/brave_rewards/browser/rewards_service_impl.cc:
```diff
@@ -884,9 +907,12 @@ void RewardsServiceImpl::OnLedgerStateLoaded(
+                                           : ledger::Result::LEDGER_OK;
+  handler->OnLedgerStateLoaded(error_code, data);
+
+  TriggerOnRewardsInitialized(error_code);
```
This doesn't look correct to me. RewardsServiceImpl should not be deciding when rewards is initialized and it shoudln't assume rewards is initialized when the ledger state is loaded. That leaks internals of the ledger library into RewardsService. This should be triggered from ledger_impl.

bridiver replied:
I know it isn't part of this PR, but we shouldn't be calling `GetRewardsMainEnabled` below either. The LedgerClient methods in particular should not be doing anything other than handling the load/callback.

bridiver replied:
and actually I think this is redundant and you should be using `OnWalletInitialized`

emerick replied:
Oh, you're right. For some reason, I was thinking that was only called when the wallet was initially created. I'll fix that, thanks for noticing!

bridiver commented on components/services/bat_ledger/public/interfaces/bat_ledger.mojom:
```diff
@@ -100,6 +100,8 @@ interface BatLedger {
 
   SetCatalogIssuers(string info);
   AdSustained(string info);
+
```
nit: why the extra space here?

emerick replied:
It looked like vertical whitespace was being used in that file to group similar functions, but I'll remove it.

bridiver commented on patches/chrome-browser-ui-webui-version_ui.cc.patch:
```diff
@@ -0,0 +1,21 @@
+       WebUIDataSource::Create(chrome::kChromeUIVersionHost);
+ 
+   // Localized and data strings.
++  html_source->AddLocalizedString(version_ui::kBraveRewardsDetails,
```
we shouldn't be patching here for this. cc @petemill for a better place to put this

bridiver replied:
actually I don't think this should be needed at all. Whatever version info we need can go in the internals ui page itself or in the rewards ui

emerick replied:
I'm totally happy to move this, but it was specifically requested to be here in https://github.com/brave/brave-browser/issues/1174#issuecomment-445906938.

bridiver commented on patches/components-version_ui-version_ui_constants.cc.patch:
```diff
@@ -0,0 +1,11 @@
+ const char kVersion[] = "version";
+ const char kVersionBitSize[] = "version_bitsize";
+ const char kVersionModifier[] = "version_modifier";
++const char kBraveRewardsDetails[] = "brave_rewards_details";
```
this is also not the right place for this and won't be necessary when the above patch is replaced

bridiver replied:
if we are keeping the patch this should be done in a chromium_src override

bridiver commented on vendor/bat-native-ledger/src/bat/ledger/ledger.cc:
```diff
@@ -677,6 +677,96 @@ bool AutoContributeProps::loadFromJson(const std::string& json) {
   return !error;
 }
 
+CurrentReconcileInfo::CurrentReconcileInfo()
```
nit: why is this `CurrentReconcileInfo` and not just `ReconcileInfo`?

emerick replied:
Yeah, that's wrong. I'll fix.

bridiver commented on patches/components-version_ui-resources-about_version.html.patch:
```diff
@@ -0,0 +1,15 @@
+ </if>
+         <div id="company">$i18n{company}</div>
+         <div id="copyright">$i18n{copyright}</div>
++        <hr>
```
I don't think we need to add anything here for an internal ui used for debugging. That just doesn't seem like it would justify a patch

emerick replied:
See https://github.com/brave/brave-browser/issues/1174#issuecomment-445906938 and other comments in that issue for how this link came to be. Totally happy to move it, but it seemed like this was the only choice that seemed to satisfy everyone while also minimizing patching.

bridiver replied:
@petemill if we are forced to keep this is there a better way to do it?

emerick replied:
Just so this is documented here: We're going to remove the link from chrome://version and brave/brave-browser#3538 was entered to track where it should ultimately end up.

bridiver commented on components/brave_rewards/resources/BUILD.gn:
```diff
@@ -7,16 +7,25 @@ import("//tools/grit/repack.gni")
   ]
 }
 
+pack_web_resources("internals_generated_resources") {
```
does this really need to be separate from the rest of the UI resources? Seems to just complicate things. Is there any benefit?

emerick replied:
I think I was just following the existing example in rewards (I'll have to check), but I can move it.

bridiver replied:
if there is no specific reason then I don't think we should add another pak file here. They are generally per-component, not per-webui

emerick replied:
Sure, I'll eliminate it.

emerick commented:
> page doesn't update

@NejcZdovc Good point, fixed!

emerick commented:
Addressed most of the review feedback you had @bridiver, thanks! The main thing I didn't address in your feedback is the addition of the chrome://rewards-internals link to chrome://version. I think we probably need input from the product folks and tech support if we want to move/adjust that (it was their request to place a link to that page in a central location with other "tech support" info, so I just want to make sure we don't inadvertently skip that requirement).

NejcZdovc commented:
@emerick I would leave link in as this was requirement from product/QA. Can you please rebase and we can merge it

emerick commented:
@bridiver This is ready to merge, just wanted to make sure you're comfortable with the rewards link in chrome://version. I think I addressed your other feedback, let me know what you think.

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,128 @@
+  DCHECK(rewards_service_);
+  rewards_service_->GetRewardsInternalsInfo(
+      base::Bind(&BraveRewardsInternalsUI::OnGetRewardsInternalsInfo,
+                 base::Unretained(this)));
```
this isn't safe because rewards service can outlive the webui page. You need WeakPtr here

emerick replied:
Thanks for catching.

bridiver commented on components/brave_rewards/browser/rewards_service_impl.cc:
```diff
@@ -793,6 +793,29 @@ void RewardsServiceImpl::OnGetAutoContributeProps(
+  info.loadFromJson(json_info);
+
+  auto rewards_internals_info =
+    std::make_unique<brave_rewards::RewardsInternalsInfo>();
```
lint: this should be 4 spaces

bridiver commented on components/brave_rewards/browser/rewards_service_impl.cc:
```diff
@@ -2444,6 +2468,12 @@ void RewardsServiceImpl::SetBackupCompleted() {
+void RewardsServiceImpl::GetRewardsInternalsInfo(
+    const GetRewardsInternalsInfoCallback& callback) {
+  bat_ledger_->GetRewardsInternalsInfo(base::BindOnce(
+      &RewardsServiceImpl::OnGetRewardsInternalsInfo, AsWeakPtr(), callback));
```
this actually doesn't need to use WeakPtr because `RewardsServiceImpl` will always outlive `bat_ledger_`, but it doesn't hurt anything. I just want to make sure everyone understands the issues with callbacks and object lifetimes

bridiver commented on patches/components-version_ui-version_ui_constants.h.patch:
```diff
@@ -0,0 +1,12 @@
+ extern const char kVersion[];
+ extern const char kVersionBitSize[];
+ extern const char kVersionModifier[];
++extern const char kBraveRewardsDetails[];
```
see above

bridiver commented on vendor/bat-native-ledger/BUILD.gn:
```diff
@@ -72,6 +72,7 @@ source_set("headers") {
     "include/bat/ledger/media_publisher_info.h",
     "include/bat/ledger/pending_contribution.h",
     "include/bat/ledger/publisher_info.h",
+    "include/bat/ledger/reconcile_info.h",
```
I don't see this file in the PR. Am I missing something?

bridiver commented on vendor/bat-native-ledger/include/bat/ledger/rewards_internals_info.h:
```diff
@@ -0,0 +1,32 @@
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef BAT_LEDGER_REWARDS_INTERNALS_INFO_
```
lint: these should all end in `H_` -> `BAT_LEDGER_REWARDS_INTERNALS_INFO_H_`

bridiver commented on vendor/bat-native-ledger/src/ledger_impl.cc:
```diff
@@ -7,6 +7,7 @@
 #include <iostream>
 #include <random>
 #include <sstream>
+#include <utility>
```
utility is listed twice

bridiver commented on vendor/bat-native-ledger/include/bat/ledger/ledger.h:
```diff
@@ -277,6 +278,7 @@ class LEDGER_EXPORT Ledger {
       const uint64_t from_timestamp_seconds,
       const uint64_t to_timestamp_seconds,
       ledger::ConfirmationsHistoryCallback callback) = 0;
+  virtual void GetRewardsInternalsInfo(ledger::RewardsInternalsInfo* info) = 0;
```
it seems very odd to me that we would call this `RewardsInternals` inside ledger. I'm not going to block on it, but I don't think we shouldn't be tightly coupling Brave-specific UI to ledger methods. Isn't this data already available from other ledger datatypes?

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,129 @@
+      }
+
+      std::string json;
+      base::JSONWriter::Write(current_reconciles, &json);
```
you need to check the return value here to ensure that it succeeds

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,176 @@
+void RewardsInternalsDOMHandler::HandleRefresh(const base::ListValue* args) {
+  if (!rewards_internals_ui_)
+    return;
+  rewards_internals_ui_->UpdateWebUIProperties();
```
I'm confused, this doesn't call `GetRewardsInternalsInfo` so nothing is ever updated. You need a RegisterMessageCallback for `brave_rewards_internals.getRewardsInternals` and the handler should call `GetRewardsInternalsInfo`. The `OnGetRewardsInternalsInfo` callback should call `CallJavascriptFunctionUnsafe` to send the data back to the page

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,176 @@
+
+void RewardsInternalsDOMHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "brave_rewards_internals.refresh",
```
I think this should just be replaced per the comment below

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+void BraveRewardsInternalsUI::OnWalletInitialized(
```
this doesn't make sense to me. `OnWalletInitialized` may have been called long before the page is open and then it would never get the data. I think this also just needs to be dropped in favor of the method I outlined above.

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,146 @@
+}
+
+void RewardsInternalsDOMHandler::OnPreferenceChanged() {
+  rewards_service_->GetRewardsInternalsInfo(
```
do we need to refetch this when `kBraveRewardsEnabled`? Either way I think we should split out `isRewardsEnabled` from `RewardsInternalsInfo`. Looks much better otherwise

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,146 @@
+void RewardsInternalsDOMHandler::HandleGetRewardsInternalsInfo(
+    const base::ListValue* args) {
+  rewards_service_->GetRewardsInternalsInfo(
+      base::Bind(&RewardsInternalsDOMHandler::OnGetRewardsInternalsInfo,
```
this should be `BindOnce`. You will also have to update the method for `CallbackOnce`

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,146 @@
+  pref_change_registrar_->Init(prefs);
+  pref_change_registrar_->Add(
+      brave_rewards::prefs::kBraveRewardsEnabled,
+      base::Bind(&RewardsInternalsDOMHandler::OnPreferenceChanged,
```
this should be `BindRepeating`

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,162 @@
+  if (!web_ui()->CanCallJavascript())
+    return;
+  base::DictionaryValue info_dict;
+  info_dict.SetBoolean("isRewardsEnabled", IsRewardsEnabled());
```
forgot to remove this?

emerick replied:
Oops, yes.

bridiver commented on browser/ui/webui/brave_rewards_internals_ui.cc:
```diff
@@ -0,0 +1,162 @@
+void RewardsInternalsDOMHandler::RegisterMessages() {
+  web_ui()->RegisterMessageCallback(
+      "brave_rewards_internals.getRewardsEnabled",
+      base::Bind(&RewardsInternalsDOMHandler::HandleGetRewardsEnabled,
```
sorry, this should also be Repeating. `base::Bind` is being deprecated so we should use Once/Repeating

NejcZdovc commented on vendor/bat-native-ledger/src/bat_state.cc:
```diff
@@ -6,6 +6,10 @@
 
 #include "bat_state.h"
+
+#include <algorithm>
```
duplicated `algorithm`
